File: Data Structures & Algorithms/balanced-binary-tree/submission-3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        if not root:
            return True

        def isBalancedHelper(root: Optional[TreeNode]):
            if not root:
                return 0
            
            return 1 + max(isBalancedHelper(root.left), isBalancedHelper(root.right))

        leftBalanced = self.isBalanced(root.left)
        rightBalanced = self.isBalanced(root.right)

        if (not (leftBalanced and rightBalanced)):
            return False

        leftHeight = isBalancedHelper(root.left)
        rightHeight = isBalancedHelper(root.right)

        if (abs(leftHeight - rightHeight) > 1):
            return False
        else:
            return True

File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-3.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    pair<int, int> numNodesTree(TreeNode* root, int k, int progress) {
        if (!root) return {progress, -1};

        auto [count, val] = numNodesTree(root->left, k, progress);

        //cout << "root val: " << root->val << " numReturnedLeft: " << count << endl;
        //cout << "root val: " << root->val << " progress: " << progress << endl;

        if (val != -1) {
            return {0, val};
        } else if (count + 1 == k) {
            return {0, root->val};
        }

        auto [countTwo, valTwo] = numNodesTree(root->right, k, count + 1);

        //cout << "root val: " << root->val << " numReturnedRight: " << countTwo << endl;

        if (valTwo != -1) {
            return {0, valTwo};
        }

        return {countTwo, -1};
    }

    int kthSmallest(TreeNode* root, int k) {
        auto [count, node] = numNodesTree(root, k, 0);

        return node;
    }
};
